nav2_controller/src/nav2_controller_client.cpp

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) 2019 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nav2_controller/nav2_controller_client.hpp"
+
+#include <cmath>
+#include <memory>
+
+namespace nav2_controller
+{
+
+Nav2ControllerClient::Nav2ControllerClient()
+{
+  // Create the node to use for all of the service clients
+  node_ = std::make_shared<rclcpp::Node>("nav2_controller_service_client");
+
+  // All of the services use the same (Empty) request
+  request_ = std::make_shared<Empty::Request>();
+
+  // Create the service clients
+  startup_client_ = node_->create_client<Empty>("nav2_controller/startup");
+  pause_client_ = node_->create_client<Empty>("nav2_controller/pause");
+  resume_client_ = node_->create_client<Empty>("nav2_controller/resume");
+  shutdown_client_ = node_->create_client<Empty>("nav2_controller/shutdown");
+
+  navigate_action_client_ =
+    rclcpp_action::create_client<nav2_msgs::action::NavigateToPose>(node_, "NavigateToPose");
+
+  initial_pose_publisher_ =
+    node_->create_publisher<geometry_msgs::msg::PoseWithCovarianceStamped>("initialpose");
+}
+
+void
+Nav2ControllerClient::startup()
+{
+  callService(startup_client_, "nav2_controller/startup");
+}
+
+void
+Nav2ControllerClient::shutdown()
+{
+  callService(shutdown_client_, "nav2_controller/shutdown");
+}
+
+void
+Nav2ControllerClient::pause()
+{
+  callService(pause_client_, "nav2_controller/pause");
+}
+
+void
+Nav2ControllerClient::resume()
+{
+  callService(resume_client_, "nav2_controller/resume");
+}
+
+geometry_msgs::msg::Quaternion
+Nav2ControllerClient::orientationAroundZAxis(double angle)
+{
+  auto orientation = geometry_msgs::msg::Quaternion();
+
+  orientation.x = 0.0;
+  orientation.y = 0.0;
+  orientation.z = sin(angle) / (2 * cos(angle / 2));
+  orientation.w = cos(angle / 2);
+
+  return orientation;
+}
+
+void
+Nav2ControllerClient::set_initial_pose(double x, double y, double theta)
+{
+  const double PI = 3.141592653589793238463;
+  geometry_msgs::msg::PoseWithCovarianceStamped pose;
+
+  pose.header.frame_id = "map";
+  pose.header.stamp = node_->now();
+  pose.pose.pose.position.x = x;
+  pose.pose.pose.position.y = y;
+  pose.pose.pose.position.z = 0.0;
+  pose.pose.pose.orientation = orientationAroundZAxis(theta);
```

Comment:
```suggestion
  pose.pose.pose.orientation = tf2::Quaternion().setRPY(0, 0, theta).toMsg();
```

I'm not sure if the toMsg is necessary. It might get done automatically. 

Alternatively, you could just do this in the body of your function instead of calculating it manually, while still keeping the nice naming.

Reply:
I've changed to code to match the suggestion as closely as I could. setRPY returns void and toMsg is a separate function, so it had to be slightly different.